src/vm_cpu.rs

```diff
@@ -149,6 +160,120 @@ impl VmPmuState {
+            0xC18 => Some(read_csr::<0xC18>()),
+            0xC19 => Some(read_csr::<0xC19>()),
+            0xC1A => Some(read_csr::<0xC1A>()),
+            0xC1B => Some(read_csr::<0xC1B>()),
+            0xC1C => Some(read_csr::<0xC1C>()),
+            0xC1D => Some(read_csr::<0xC1D>()),
+            0xC1E => Some(read_csr::<0xC1E>()),
+            0xC1F => Some(read_csr::<0xC1F>()),
+            _ => None,
+        }
+    }
+
+    /// Captures the current CSR value of started counters, after stopping them.
+    /// This is intended to be called in anticipation of out-bound context switch.
+    #[allow(dead_code)]
+    pub fn save_counters(&mut self) {
+        use drivers::pmu::PmuInfo;
+        use sbi::api::pmu;
+        let pmu_info = PmuInfo::get();
+        for counter_index in
+            self.started_counters.trailing_zeros() as u64..pmu_info.get_num_counters()
+        {
+            if self.started_counters & (1 << counter_index) != 0 {
+                let counter_index_mask = Self::gen_counter_mask(counter_index, counter_index);
+                // Ignore errors since we can't do anything about them.
+                let _ = pmu::stop_counters(
+                    counter_index,
+                    counter_index_mask,
+                    PmuCounterStopFlags::default(),
+                );
+                if let Some(counter_info) = pmu_info.get_counter_info(counter_index) {
+                    if counter_info.is_hardware_counter() {
+                        if let Some(value) = Self::read_pmu_csr(counter_info.get_csr()) {
+                            self.counter_values[counter_index as usize] = value;
+                        }
+                    } else if let Ok(value) = pmu::read_firmware_counter(counter_index) {
+                        self.counter_values[counter_index as usize] = value;
+                    }
+                }
+            }
+        }
+    }
+
+    /// Starts previously started counters using the last captured value as the initial value.
+    /// This is intended to be called in of anticipation of an in-bound context switch.
+    #[allow(dead_code)]
+    pub fn restore_counters(&mut self) {
+        use drivers::pmu::PmuInfo;
+        use sbi::api::pmu;
+        let pmu_info = PmuInfo::get();
+        for counter_index in
+            self.configured_counters.trailing_zeros() as u64..pmu_info.get_num_counters()
+        {
+            let event_type =
+                PmuEventType::from_raw_value(self.configured_counter_map[counter_index as usize]);
+            let counter_info = pmu_info.get_counter_info(counter_index);
+            if self.configured_counters & (1 << counter_index) != 0 && let Ok(event_type) = event_type  && counter_info.is_some() {
+                let counter_index_mask = Self::gen_counter_mask(counter_index, counter_index);
+                let config_flags = PmuCounterConfigFlags::default().set_skip_match().set_sinh().set_minh();
```

Comment:
Sort of related to the above, but based on my read of the implementation in OpenSBI the `SKIP_MATCH` bit doesn't result in the counter getting re-configured; it just auto-starts or clears the counter based on the other bits.

Reply:
Hmm...my read was that it simply skips the wild card matching, but will take a peek at the OpenSBI implementation.

**NOTE: When SBI_PMU_CFG_FLAG_SKIP_MATCH is set in config_flags, the SBI implementation will unconditionally select the first counter from the set of counters specified by the counter_idx_base and counter_idx_mask.**

Reply:
Yes. It also assumes that the the first counter in counter mask is a valid one and configured for that event. That sanity check should be there. Otherwise, implementation must return error.